File: android/src/main/java/com/nitroplay/support/errors/NitroPlayerError.kt
package com.nitroplay.video.core

// Base class for all video errors
abstract class NitroPlayerError(
    open val code: String,
    message: String
) : Error("{%@$code::$message@%}")

// Library related errors
sealed class LibraryError(code: String, message: String) : NitroPlayerError(code, message) {
  object Deallocated : LibraryError(
    "library/deallocated",
    "Object has been deallocated"
  )

  object ApplicationContextNotFound : LibraryError(
    "library/application-context-not-found",
    "Application context not found"
  )

  class MethodNotSupported(val methodName: String) : LibraryError(
    "library/method-not-supported",
    "Method $methodName() is not supported on Android"
  )
}

// Player related errors
sealed class PlayerError(code: String, message: String) : NitroPlayerError(code, message) {
  object NotInitialized : PlayerError(
    "player/not-initialized",
    "Player has not been initialized (Or has been set to null)"
  )

  object AssetNotInitialized : PlayerError(
    "player/asset-not-initialized",
    "Asset has not been initialized (Or has been set to null)"
  )

  object InvalidSource : PlayerError(
    "player/invalid-source",
    "Invalid source passed to player"
  )

  object Cancelled : PlayerError(
    "player/cancelled",
    "Operation was cancelled"
  )
}

// Source related errors
sealed class SourceError(code: String, message: String) : NitroPlayerError(code, message) {
  class InvalidUri(val uri: String) : SourceError(
    "source/invalid-uri",
    "Invalid source file uri: $uri"
  )

  class MissingReadFilePermission(val uri: String) : SourceError(
    "source/missing-read-file-permission",
    "Missing read file permission for source file at $uri"
  )

  class FileDoesNotExist(val uri: String) : SourceError(
    "source/file-does-not-exist",
    "File does not exist at URI: $uri"
  )

  object FailedToInitializeAsset : SourceError(
    "source/failed-to-initialize-asset",
    "Failed to initialize asset"
  )

  class UnsupportedContentType(val uri: String) : SourceError(
    "source/unsupported-content-type",
    "type of content (${uri}) is not supported"
  )

  object Cancelled : SourceError(
    "source/cancelled",
    "Operation was cancelled"
  )
}

// View related errors
sealed class NitroPlayerViewError(code: String, message: String) : NitroPlayerError(code, message) {
  class ViewNotFound(val viewId: Int) : NitroPlayerViewError(
    "view/not-found",
    "View with viewId $viewId not found"
  )
  object ViewIsDeallocated : NitroPlayerViewError(
    "view/deallocated",
    "Attempt to access a view, but it has been deallocated (or not initialized)"
  )
}

// Unknown error
class UnknownError : NitroPlayerError("unknown/unknown", "Unknown error")

File: ios/support/errors/NitroPlayerError.swift
//
//  NitroPlayerFileHelper.swift
//  NitroPlay
//
//  Created by Krzysztof Moch on 24/01/2025.
//

import Foundation
import NitroModules

// MARK: - LibraryError
enum LibraryError: NitroPlayerError {
  case deallocated(objectName: String)

  var code: String {
    switch self {
    case .deallocated:
      return "library/deallocated"
    }
  }

  var message: String {
    switch self {
    case let .deallocated(objectName: objectName):
      return "Object \(objectName) has been deallocated"
    }
  }
}

// MARK: - PlayerError
enum PlayerError: NitroPlayerError {
  case notInitialized
  case assetNotInitialized
  case invalidSource
  case invalidTrackUrl(url: String)
  case cancelled
  
  var code: String {
    switch self {
    case .notInitialized:
      return "player/not-initialized"
    case .assetNotInitialized:
      return "player/asset-not-initialized"
    case .invalidSource:
      return "player/invalid-source"
    case .invalidTrackUrl:
      return "player/invalid-track-url"
    case .cancelled:
      return "player/cancelled"
    }
  }
  
  var message: String {
    switch self {
    case .notInitialized:
      return "Player has not been initialized (Or has been set to nil)"
    case .assetNotInitialized:
      return "Asset has not been initialized (Or has been set to nil)"
    case .invalidSource:
      return "Invalid source passed to player"
    case let .invalidTrackUrl(url: url):
      return "Invalid track URL: \(url)"
    case .cancelled:
      return "Operation was cancelled"
    }
  }
}

// MARK: - SourceError
enum SourceError: NitroPlayerError {
  case invalidUri(uri: String)
  case missingReadFilePermission(uri: String)
  case fileDoesNotExist(uri: String)
  case failedToInitializeAsset
  case unsupportedContentType(uri: String)
  case cancelled

  var code: String {
    switch self {
    case .invalidUri:
      return "source/invalid-uri"
    case .missingReadFilePermission:
      return "source/missing-read-file-permission"
    case .fileDoesNotExist:
      return "source/file-does-not-exist"
    case .failedToInitializeAsset:
      return "source/failed-to-initialize-asset"
    case .unsupportedContentType:
      return "source/unsupported-content-type"
    case .cancelled:
      return "source/cancelled"
    }
  }

  var message: String {
    switch self {
    case let .invalidUri(uri: uri):
      return "Invalid source file uri: \(uri)"
    case let .missingReadFilePermission(uri: uri):
      return "Missing read file permission for source file at \(uri)"
    case let .fileDoesNotExist(uri: uri):
      return "File does not exist at URI: \(uri)"
    case .failedToInitializeAsset:
      return "Failed to initialize asset"
    case let .unsupportedContentType(uri: uri):
      return "type of content (\(uri)) is not supported"
    case .cancelled:
      return "Operation was cancelled"
    }
  }
}

// MARK: - NitroPlayerViewError
enum NitroPlayerViewError: NitroPlayerError {
  case viewNotFound(nitroId: Double)
  case viewIsDeallocated

  var code: String {
    switch self {
    case .viewNotFound:
      return "view/not-found"
    case .viewIsDeallocated:
      return "view/deallocated"
    }
  }

  var message: String {
    switch self {
    case let .viewNotFound(nitroId: nitroId):
      return "View with nitroId \(nitroId) not found"
    case .viewIsDeallocated:
      return "Attempt to access a view, but it has been deallocated (or not initialized)"
    }
  }
}

// MARK: - UnknownError
struct UnknownError: NitroPlayerError {
  var code: String { "unknown/unknown" }
  var message: String { "Unknown error" }
}

// MARK: - NitroPlayerError
protocol NitroPlayerError {
  var code: String { get }
  var message: String { get }
}

extension NitroPlayerError {
  private func getMessage() -> String {
    return "{%@\(code)::\(message)@%}"
  }

  func error() -> Error {
    return RuntimeError.error(withMessage: getMessage())
  }
}

